pkg/registry/service/rest.go

```diff
@@ -265,6 +276,97 @@ func (*REST) NewList() runtime.Object {
+					apiservice.AnnotationHealthCheckNodePort, errmsg)}
+				return false, errors.NewInvalid(api.Kind("Service"), service.Name, el)
+			}
+			glog.Infof("Reserved user requested nodePort: %d", requestedHealthCheckNodePort)
+		} else {
+			// If the request has no health check nodePort specified, allocate any
+			healthCheckNodePort, err := rs.serviceNodePorts.AllocateNext()
+			if err != nil {
+				// TODO: what error should be returned here?  It's not a
+				// field-level validation failure (the field is valid), and it's
+				// not really an internal error.
+				return false, errors.NewInternalError(fmt.Errorf("failed to allocate a nodePort: %v", err))
+			}
+			// Insert the newly allocated health check port as an annotation (plan of record for Alpha)
+			service.Annotations[apiservice.AnnotationHealthCheckNodePort] = fmt.Sprintf("%d", healthCheckNodePort)
+			glog.Infof("Reserved health check nodePort: %d", healthCheckNodePort)
+		}
+
+	case oldServiceHasHealthCheckNodePort && !assignHealthCheckNodePort:
+		glog.Infof("Transition from OnlyLocal LB service to Global service")
+		err := rs.serviceNodePorts.Release(int(oldHealthCheckNodePort))
+		if err != nil {
+			glog.Warningf("Error releasing service health check %s node port %d: %v", service.Name, oldHealthCheckNodePort, err)
+			return false, errors.NewInternalError(fmt.Errorf("failed to free health check nodePort: %v", err))
+		} else {
+			delete(service.Annotations, apiservice.AnnotationHealthCheckNodePort)
+			glog.Infof("Freed health check nodePort: %d", oldHealthCheckNodePort)
+		}
+
+	case !oldServiceHasHealthCheckNodePort && !assignHealthCheckNodePort:
+		if _, ok := service.Annotations[apiservice.AnnotationHealthCheckNodePort]; ok {
+			glog.Warningf("Attempt to insert health check node port annotation DENIED")
+			el := field.ErrorList{field.Invalid(field.NewPath("metadata", "annotations"),
+				apiservice.AnnotationHealthCheckNodePort, "Cannot insert healthcheck nodePort annotation")}
+			return false, errors.NewInvalid(api.Kind("Service"), service.Name, el)
+		}
+
+	case oldServiceHasHealthCheckNodePort && assignHealthCheckNodePort:
+		if _, ok := service.Annotations[apiservice.AnnotationHealthCheckNodePort]; !ok {
+			glog.Warningf("Attempt to delete health check node port annotation DENIED")
+			el := field.ErrorList{field.Invalid(field.NewPath("metadata", "annotations"),
+				apiservice.AnnotationHealthCheckNodePort, "Cannot delete healthcheck nodePort annotation")}
+			return false, errors.NewInvalid(api.Kind("Service"), service.Name, el)
+		}
+		if oldHealthCheckNodePort != requestedHealthCheckNodePort {
+			glog.Warningf("Attempt to change value of health check node port annotation DENIED")
+			el := field.ErrorList{field.Invalid(field.NewPath("metadata", "annotations"),
+				apiservice.AnnotationHealthCheckNodePort, "Cannot change healthcheck nodePort during update")}
+			return false, errors.NewInvalid(api.Kind("Service"), service.Name, el)
+		}
+	}
```

Comment:
it's useful to add a comment here saying this is the "normal" update path, i.e the else clause is what happens if i modify labels on a Service using esipp.